full-service/src/service/password_manager.rs

```diff
@@ -0,0 +1,377 @@
+        //       string. This is ok for our use case, see discussion on precomputation attacks:
+        //       https://crypto.stackexchange.com/questions/77549/is-it-safe-to-use-a-deterministic-salt-as-an-input-to-kdf-argon2
+        let mut hasher = Blake2b::new();
+        hasher.update(&SALT_DOMAIN_TAG);
+        hasher.update(&password);
+        let salt = hasher.finalize();
+        let config = argon2::Config::default();
+        Ok(argon2::hash_raw(password.as_bytes(), &salt, &config)?)
+    }
+
+    fn set_password(&self, password: String) -> Result<bool, PasswordServiceError> {
+        let password_hash = self.get_password_hash(password)?;
+
+        let conn_context = self.wallet_db.get_conn_context()?;
+        conn_context.conn.transaction::<(), PasswordServiceError, _>(|| {
+            match EncryptionIndicator::get_encryption_state(&conn_context.conn)? {
+                EncryptionState::Empty => {
+                    log::info!(
+                        self.logger,
+                        "Database has never been locked and has no accounts. Setting password for future accounts."
+                    );
+                    self.wallet_db.set_password_hash(&password_hash, &conn_context.conn)?;
+                }
+                EncryptionState::Encrypted => {
+                    return Err(PasswordServiceError::DatabaseEncrypted);
+                }
+                EncryptionState::Unencrypted => {
+                    log::info!(
+                        self.logger,
+                        "Database is unencrypted. Setting password with new password, and encrypting all accounts."
+                    );
+                    self.wallet_db.set_password_hash(&password_hash, &conn_context.conn)?;
+                    for account in Account::list_all(&conn_context.conn)? {
+                        let encrypted_account_key = conn_context.encryption_provider.encrypt(
+                            &account.account_key,
+                        )?;
+                        account.update_encrypted_account_key(&encrypted_account_key, &conn_context.conn)?;
+                    }
+                }
+            }
+            Ok(())
+        })?;
+        Ok(true)
+    }
+
+    fn unlock(&self, password: String) -> Result<bool, PasswordServiceError> {
+        let password_hash = self.get_password_hash(password)?;
+        self.wallet_db.unlock(&password_hash)?;
+        Ok(true)
+    }
+
+    fn change_password(
+        &self,
+        old_password: String,
+        new_password: String,
+    ) -> Result<bool, PasswordServiceError> {
+        // Make sure DB is unlocked to continue.
+        self.unlock(old_password.clone())?;
+
+        let old_password_hash = self.get_password_hash(old_password)?;
+        let new_password_hash = self.get_password_hash(new_password)?;
+
+        // Re-encrypt all of our accounts with the new password hash
```

Comment:
Having to read, decrypt, encrypt, and write every record in the DB won't scale well. That may be OK for accounts (assuming that a wallet will only have one or a few accounts), but this approach probably wouldn't be a good fit if we want to later do encryption-at-rest for larger wallet datasets, like received outputs.

I think it's more standard to use a data-encryption-key (DEK) and a key-encryption-key (KEK) to solve this problem (aka "envelope encryption"), where the "master" KEK is used to decrypt the DEK, and then the DEK is used to encrypt/decrypt each piece of stored data. Then, changing the password is equivalent to decrypting the DEK using the old KEK and re-encrypting it using the new KEK (which is derived from the new password). The DEK doesn't change, so it is not necessary to re-encrypt each piece of stored data.